src/D2L.Security.OAuth2/Provisioning/Default/AuthServiceClient.cs

```diff
@@ -45,17 +46,60 @@ Uri authEndpoint
 		/// <param name="assertion">A JWT signed by the private key of the entity requesting the token</param>
 		/// <param name="scopes">List of scopes to include in the access token</param>
 		/// <returns>A JWT token from the auth service signed with the auth service's private key</returns>
+		/// <exception cref="AuthServiceException">
+		/// The auth service could not be reached, or it did not respond with
+		/// a status code indicating success.
+		/// </exception>
 		async Task<IAccessToken> IAuthServiceClient.ProvisionAccessTokenAsync(
 			string assertion,
 			IEnumerable<Scope> scopes
 		) {
 			string requestBody = BuildFormContents( assertion, scopes );
-			using( HttpResponseMessage response = await MakeRequest( requestBody ).SafeAsync() ) {
-				response.EnsureSuccessStatusCode();
-
-				using( var resultStream = await response.Content.ReadAsStreamAsync().SafeAsync() ) {
-					IAccessToken accessToken = SerializationHelper.ExtractAccessToken( resultStream );
-					return accessToken;
+			HttpResponseMessage response = null;
+			try {
+				try {
+					response = await MakeRequest( requestBody ).SafeAsync();
+				} catch( TaskCanceledException exception ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.Timeout,
+						message: "The web request to the Auth Service has timed out.",
+						innerException: exception
+					);
+				} catch( Exception exception ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.ConnectionFailure,
+						message: "Could not establish a connection with the Auth Service.",
+						innerException: exception
+					);
+				}
+				
+				if( !response.IsSuccessStatusCode ) {
+					throw new AuthServiceException(
+						errorType: ServiceErrorType.ErrorResponse,
+						serviceStatusCode: response.StatusCode,
+						message: string.Concat(
+							"The Auth Service did not respond with a status code indicating success. ",
+							"(Received ", ((int)response.StatusCode).ToString(), " ",
+							response.ReasonPhrase, ")"

```

Comment:
Interesting idea... `AuthServiceException` is pretty important so it doesn't seem 100% wrong to have its own handler. I do get a bit worried about an explosion in the LE but I also like the type-y nature (rather than just jamming more shit into a string.)

I'll defer to you to figure out what's best.